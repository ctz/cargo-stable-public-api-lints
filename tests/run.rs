use std::mem;
use std::process::{Command, Output};

#[test]
fn non_exhaustive_enum_fixed() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-enum",
        "enum_fixed"
    ));
}

#[test]
fn non_exhaustive_enum_warn() {
    insta::assert_debug_snapshot!(run("tests/non_exhaustive/", "exhaustive-enum", "enum_warn"));
}

#[test]
fn non_exhaustive_enum_allow() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-enum",
        "enum_allow"
    ));
}

#[test]
fn non_exhaustive_struct_fixed() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-struct",
        "struct_fixed"
    ));
}

#[test]
fn non_exhaustive_struct_warn() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-struct",
        "struct_warn"
    ));
}

#[test]
fn non_exhaustive_struct_allow() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-struct",
        "struct_allow"
    ));
}

#[test]
fn drop_struct_warn() {
    insta::assert_debug_snapshot!(run("tests/drop/", "drop", "struct_warn"));
}

#[test]
fn sorted_enum() {
    insta::assert_debug_snapshot!(run("tests/sorted_enum/", "sorted-enum", "default"));
}

fn run(path: &str, lint_id: &str, feature: &str) -> Output {
    let mut output = Command::new("cargo")
        .arg("-q")
        .arg("run")
        .arg("--")
        .arg("--features")
        .arg(feature)
        .arg("--allow")
        .arg("all")
        .arg("--warn")
        .arg(lint_id)
        .arg(format!("{path}/Cargo.toml"))
        .output()
        .unwrap();

    // currently the rustdoc subprocess writes unstable output
    // to stderr
    println!(
        "stderr: {}",
        String::from_utf8(mem::take(&mut output.stderr)).expect("stderr is not valid utf8")
    );
    println!(
        "stdout:\n{}",
        String::from_utf8(output.stdout.clone()).expect("stdout is not valid utf8")
    );
    output
}
