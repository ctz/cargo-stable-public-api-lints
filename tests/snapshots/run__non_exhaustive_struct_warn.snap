---
source: tests/run.rs
expression: "run(\"tests/non_exhaustive/\", \"exhaustive-struct\", \"struct_warn\")"
---
Output {
    status: ExitStatus(
        unix_wait_status(
            256,
        ),
    ),
    stdout: "------------------------------------------------------------\nMarking a struct `#[non_exhaustive]` means future changes can add new fields.\n\nAnnotate this type with `#[cfg_attr(spal, allow(spal::exhaustive_struct))]` to ignore this lint\n------------------------------------------------------------\n\n⚠\u{fe0f} struct `pub struct non_exhaustive::struct_warn::NormalEmpty` should be marked #[non_exhaustive]\n  at src/struct_warn.rs:13\n\n⚠\u{fe0f} struct `pub struct non_exhaustive::struct_warn::NormalPublic` should be marked #[non_exhaustive]\n  at src/struct_warn.rs:14\n\n⚠\u{fe0f} struct `pub struct non_exhaustive::struct_warn::TupleEmpty()` should be marked #[non_exhaustive]\n  at src/struct_warn.rs:8\n\n⚠\u{fe0f} struct `pub struct non_exhaustive::struct_warn::TuplePublic(pub usize, pub usize)` should be marked #[non_exhaustive]\n  at src/struct_warn.rs:10\n\n⚠\u{fe0f} struct `pub struct non_exhaustive::struct_warn::Unit` should be marked #[non_exhaustive]\n  at src/struct_warn.rs:6\n\n",
    stderr: "",
}
